### PR TITLE
Add 'opensearch' and 'opensearch_info' modules

### LINKED ISSUE #858: OpenSearch, ElasticSearch module
### Summary

It would be useful to have a `ec2_opensearch` module to deploy opensearch/elasticsearch clusters. Is it missing because nobody from the community submitted a PR? Or was it intentionally not added? Some other reason?

There is one module outside this repo (https://github.com/fiunchinho/ansible-aws-elasticsearch-module), but it hasn't been updated in the last 3 years.

### Issue Type

Feature Idea

### Component Name

opensearch

### Additional Information

```yaml (paste below)

```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

### PR DESCRIPTION
**Pull request by sebastien-rosset**
##### SUMMARY

1. Add `opensearch` module to create/update AWS OpenSearch/Elasticsearch  domains.
2. Add `opensearch_info` module to query AWS OpenSearch/Elasticsearch domains.

Fixes #858

Requires https://github.com/mattclay/aws-terminator/pull/187

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

Creates OpenSearch or ElasticSearch domain.

##### ADDITIONAL INFORMATION

1. The minimum version of `botocore` for these modules is 1.21.38. 
2. The integration tests take more than 4 hours to execute. Tests time out in the CI. I was able to run the integration tests locally.

```paste below

```

### CODE REVIEW

**Comment by markuman** [comment]
It does not come to the integration tests, because of some sanity failures.  
You can run it easily locally

` ansible-test sanity --python 3.8 plugins/modules/opensearch.py --docker default`

**Comment by sebastien-rosset** [comment]
@alinabuzachis , besides the module itself and the integration tests, are there other files I need to add?

**Comment by sebastien-rosset** [comment]
The error below seems unrelated to my changes:

```
Run command: /root/.ansible/test/venv/sanity.yamllint/3.8/8bc751e2/bin/python /root/ansible/test/lib/ansible_test/_util/controller/tools/yamlcheck.py
Run command: /root/.ansible/test/venv/sanity.yamllint/3.8/8bc751e2/bin/python /root/ansible/test/lib/ansible_test/_util/controller/sanity/yamllint/yamllinter.py
WARNING: There following sanity test virtual environments are out-of-date in the "default" container: sanity-3.8, sanity.ansible-doc-3.8, sanity.changelog-3.8, sanity.import-3.8, sanity.pep8-3.8, sanity.pylint-3.8, sanity.runtime-metadata-3.8, sanity.validate-modules-3.8, sanity.yamllint-3.8
ERROR: The 5 sanity test(s) listed below (out of 32) failed. See error output above for details.
ansible-doc
import --python 3.8
pep8
pylint
validate-modules
```

File: plugins/modules/opensearch.py [hunk 1]
```diff
@@ -0,0 +1,1194 @@
...
+from ansible_collections.amazon.aws.plugins.module_utils.ec2 import (
+    boto3_tag_list_to_ansible_dict,
+    ansible_dict_to_boto3_tag_list,
+)
+
+
+def get_domain(client, module, domain_name):
+    try:
+        response = client.describe_domain(DomainName=domain_name)
+    except client.exceptions.ResourceNotFoundException:
```
Comment (alinabuzachis) [inline plugins/modules/opensearch.py:447]: Import `from ansible_collections.amazon.aws.plugins.module_utils.core import is_boto3_error_code`

You can use something like this instead (even for the other cases)

```suggestion
    except is_boto3_error_code('ResourceNotFoundException'):
```

File: plugins/modules/opensearch.py [hunk 2]
```diff
@@ -0,0 +1,1194 @@
...
+        return None
+    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
```
Comment (alinabuzachis) [inline plugins/modules/opensearch.py:449]: And then add

```suggestion
    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:  # pylint: disable=duplicate-except
```

File: plugins/modules/opensearch.py [hunk 3]
```diff
@@ -0,0 +1,1194 @@
...
+    try:
+        wait_for_domain_status(client, module, domain_name, "domain_deleted")
+        return dict(changed=changed)
+    except client.exceptions.ResourceNotFoundException:
+        return dict(changed=changed)
+    except (botocore.exceptions.ClientError, botocore.exceptions.BotoCoreError) as e:
+        module.fail_json_aws(e, "awaiting domain deletion")
+
+
+def ensure_tags(client, module, resource_arn, existing_tags, tags, purge_tags):
```
Comment (alinabuzachis) [inline plugins/modules/opensearch.py:527]: I would personally move all the util methods inside a dedicated file here  https://github.com/ansible-collections/community.aws/tree/main/plugins/module_utils. 

@markuman what do you think about? Does this make sense for you?

File: plugins/modules/opensearch.py [hunk 4]
```diff
@@ -0,0 +1,1194 @@
...
+    try:
+        response = client.describe_domain_config(DomainName=domain_name)
+    except client.exceptions.ResourceNotFoundException:
+        return None
+    except (botocore.exceptions.BotoCoreError, botocore.exceptions.ClientError) as e:
+        module.fail_json_aws(e, msg="Couldn't get domain {0}".format(domain_name))
+    return response["DomainConfig"]
+
+
+def domain_to_facts(client, module, domain):
```
Comment (alinabuzachis) [inline plugins/modules/opensearch.py:464]: Consider also moving this one here https://github.com/ansible-collections/community.aws/tree/main/plugins/module_utils inside the opensearch.py dedicated file. I would personally change the name of the method since it only returns the associated tags.

File: plugins/modules/opensearch.py [hunk 5]
```diff
@@ -0,0 +1,1194 @@
...
+        module.fail_json_aws(
+            e, "Couldn't get tags for domain %s" % domain["domain_name"]
+        )
+    except KeyError:
+        module.fail_json(msg=str(domain))
+
+    return camel_dict_to_snake_dict(domain, ignore_list=["Tags"])
+
+
+def wait_for_domain_status(client, module, domain_name, waiter_name):
```
Comment (alinabuzachis) [inline plugins/modules/opensearch.py:479]: Even this one should be moved.

File: plugins/modules/opensearch.py [hunk 6]
```diff
@@ -0,0 +1,1194 @@
...
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+from __future__ import absolute_import, division, print_function
+
+DOCUMENTATION = """
+---
+module: opensearch
+short_description: Creates OpenSearch or ElasticSearch domain.
+description:
+  - Creates a new Amazon OpenSearch Service domain.
+version_added: 3.0.0
```
Comment (alinabuzachis) [inline plugins/modules/opensearch.py:13]: Since community.aws 3.0.0 has already been released, you can update `version_added: 3.1.0` for now.

**Review by alinabuzachis — COMMENTED** [review]
@sebastien-rosset I left some initial review comments with some suggestions. But the PR looks really good!

**Comment by alinabuzachis** [comment]
@sebastien-rosset These are some validation errors that need to be addressed.

```
2022-01-13 07:44:14.176284 \| fedora-35 \| ERROR: Found 23 validate-modules issue(s) which need to be resolved:
--
2447 | 2022-01-13 07:44:14.178447 \| fedora-35 \| See documentation for help: https://docs.ansible.com/ansible-core/devel/dev_guide/testing/sanity/validate-modules.html
2448 | 2022-01-13 07:44:14.178502 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-choices-do-not-match-spec: Argument 'state' in argument_spec defines choices as (['present', 'absent']) but documentation defines choices as ([])
2449 | 2022-01-13 07:44:14.178510 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-default-does-not-match-spec: Argument 'enabled' in argument_spec found in cognito_options defines default as (None) but documentation defines default as (False)
2450 | 2022-01-13 07:44:14.178524 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-default-does-not-match-spec: Argument 'engine_version' in argument_spec defines default as (None) but documentation defines default as ('OpenSearch_1.1')
2451 | 2022-01-13 07:44:14.178530 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-default-does-not-match-spec: Argument 'instance_type' in argument_spec found in cluster_config defines default as (None) but documentation defines default as ('t2.small.search')
2452 | 2022-01-13 07:44:14.178599 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-default-does-not-match-spec: Argument 'state' in argument_spec defines default as ('present') but documentation defines default as (None)
2453 | 2022-01-13 07:44:14.178606 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-default-does-not-match-spec: Argument 'warm_enabled' in argument_spec found in cluster_config defines default as (None) but documentation defines default as (False)
2454 | 2022-01-13 07:44:14.178612 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-elements-mismatch: Argument 'security_groups' in argument_spec found in vpc_options specifies elements as str,but elements is not documented
2455 | 2022-01-13 07:44:14.178618 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-elements-mismatch: Argument 'subnets' in argument_spec found in vpc_options specifies elements as str,but elements is not documented
2456 | 2022-01-13 07:44:14.178624 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-missing-type: Argument 'state' in argument_spec uses default type ('str') but documentation doesn't define type
2457 | 2022-01-13 07:44:14.178631 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-required-mismatch: Argument 'enabled' in argument_spec found in ebs_options is not required, but is documented as being required
2458 | 2022-01-13 07:44:14.178636 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-required-mismatch: Argument 'volume_size' in argument_spec found in ebs_options is not required, but is documented as being required
2459 | 2022-01-13 07:44:14.178644 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-required-mismatch: Argument 'volume_type' in argument_spec found in ebs_options is not required, but is documented as being required
2460 | 2022-01-13 07:44:14.178650 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-type-does-not-match-spec: Argument 'cluster_config' in argument_spec defines type as <class 'dict'> but documentation defines type as 'dict'
2461 | 2022-01-13 07:44:14.178656 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-type-does-not-match-spec: Argument 'cognito_options' in argument_spec defines type as <class 'dict'> but documentation defines type as 'dict'
2462 | 2022-01-13 07:44:14.178662 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-type-does-not-match-spec: Argument 'cold_storage_options' in argument_spec found in cluster_config defines type as <class 'dict'> but documentation defines type as 'dict'
2463 | 2022-01-13 07:44:14.178668 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-type-does-not-match-spec: Argument 'ebs_options' in argument_spec defines type as <class 'dict'> but documentation defines type as 'dict'
2464 | 2022-01-13 07:44:14.178678 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-type-does-not-match-spec: Argument 'encryption_at_rest_options' in argument_spec defines type as <class 'dict'> but documentation defines type as 'dict'
2465 | 2022-01-13 07:44:14.178684 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-type-does-not-match-spec: Argument 'node_to_node_encryption_options' in argument_spec defines type as <class 'dict'> but documentation defines type as 'dict'
2466 | 2022-01-13 07:44:14.178690 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-type-does-not-match-spec: Argument 'snapshot_options' in argument_spec defines type as <class 'dict'> but documentation defines type as 'dict'
2467 | 2022-01-13 07:44:14.178696 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: doc-type-does-not-match-spec: Argument 'vpc_options' in argument_spec defines type as <class 'dict'> but documentation defines type as 'dict'
2468 | 2022-01-13 07:44:14.178705 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: no-log-needed: Argument 'roles_key' in argument_spec could be a secret, though doesn't have `no_log` set found in advanced_security_options -> saml
2469 | 2022-01-13 07:44:14.178716 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: no-log-needed: Argument 'subject_key' in argument_spec could be a secret, though doesn't have `no_log` set found in advanced_security_options -> saml
2470 | 2022-01-13 07:44:14.178722 \| fedora-35 \| ERROR: plugins/modules/opensearch.py:0:0: undocumented-parameter: Argument 'state' is listed in the argument_spec, but not documented in the module documentation
2471 | 2022-01-13 07:44:14.210330 \| fedora-35 \| Running sanity test "yamllint"
```

File: tests/integration/targets/opensearch/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,413 @@
...
+  #     security_token: "{{ security_token | default(omit) }}"
+  #     region: "{{ aws_region }}"
+
+  collections:
+    - amazon.aws
+
+  block:
+    - name: Create VPC for use in testing
+      ec2_vpc_net:
+        name: "{{ resource_prefix }}-vpc"
```
[the remark this answers is not in the document]
Comment (sebastien-rosset) [reply tests/integration/targets/opensearch/tasks/main.yml]: done

File: tests/integration/targets/opensearch/aliases [hunk 8]
```diff
@@ -0,0 +1,5 @@
+# reason: slow
+# Tests take more than 30 minutes
+disabled
```
[the remark this answers is not in the document]
Comment (sebastien-rosset) [reply tests/integration/targets/opensearch/aliases]: done

File: plugins/modules/opensearch.py (hunk 6 above)
Comment (sebastien-rosset) [reply plugins/modules/opensearch.py]: done

File: plugins/modules/opensearch.py (hunk 3 above)
Comment (sebastien-rosset) [reply plugins/modules/opensearch.py:533]: Is this a new pattern? I see only two modules externalize functions under module_utils. It seems this would make it harder to locally set or override the `opensearch` module. I.e. when a module has a bug or needs some enhancement, if it's self contained users can just copy the `opensearch.py` file and configure the library search path. If the functionality is split, then it makes it a bit harder.

File: tests/integration/targets/opensearch/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,413 @@
...
+
+  # module_defaults:
+  #   group/aws:
+  #     aws_access_key: "{{ aws_access_key }}"
+  #     aws_secret_key: "{{ aws_secret_key }}"
+  #     security_token: "{{ security_token | default(omit) }}"
+  #     region: "{{ aws_region }}"
+
+  block:
+    - name: test without specifying required module options
```
[the remark this answers is not in the document]
Comment (alinabuzachis) [reply tests/integration/targets/opensearch/tasks/main.yml:14]: Is the creation of the cluster the only task requiring around 30 min? If yes, I would do something like this.
1. Create a cluster with `check_mode: true`
2. Create a cluster
3. Create a cluster with check_mode: true (idempotency)
4.  Create a cluster (idempotency)
Here, I assume the cluster is created only once.

and then all the other tasks that cover the module functionality (run with and without check_mode).

Does this make sense to you?

File: plugins/modules/opensearch.py (hunk 3 above)
Comment (alinabuzachis) [reply plugins/modules/opensearch.py:533]: Many modules import module_utils hosted in amazon.aws. I suggested to do so because if we add another module (for example opensearch_info.py), we can directly import the needed functions from module_utils instead of having to rewrite them.
The main objective of module_utils is to host helpers that can be used multiple times by different modules. in this way, we write the code once and then reduce redundancy and errors.

By the way, for the split, I would wait for @markuman's confirmation.

File: tests/integration/targets/opensearch/tasks/main.yml (hunk 9 above)
[the remark this answers is not in the document]
Comment (sebastien-rosset) [reply tests/integration/targets/opensearch/tasks/main.yml:14]: yes, I am adding more tests, and I'm adding support for check_mode.

File: plugins/modules/opensearch.py (hunk 3 above)
Comment (sebastien-rosset) [reply plugins/modules/opensearch.py:531]: Are you suggesting to import from `amazon.aws` or from `community.aws`? You are referencing both. In `community.aws`, there are only two files. I'm not too keen on adding these files to `amazon.aws`, this is most likely going to make the PR process much more involved, right? Open 2 PRs, make sure both are approved, make sure they work together, and if subsequently a change needs to be made, refactors would be much harder to do.

**Comment by sebastien-rosset** [comment]
It looks like the `opensearch` service was added in boto3 versions 1.18.38, 1.18.61, 1.20.
However, the integration tests are executed with boto3 version 1.16.0. Can the version be upgraded?
```
fatal: [testhost]: FAILED! => {
    "boto3_version": "1.16.0",
    "botocore_version": "1.19.0",

```

File: requirements.txt [hunk 10]
```diff
@@ -2,7 +2,7 @@
 # - tests/unit/constraints.txt
 # - tests/integration/constraints.txt
 # - tests/integration/targets/setup_botocore_pip
-botocore>=1.19.0
-boto3>=1.16.0
+botocore>=1.23.0
```
Comment (sebastien-rosset) [inline requirements.txt:5]: See PR comments. I had to do this for the test to run. boto3 version 1.16 does not have the `opensearch` service.

**Review by tremble — CHANGES_REQUESTED** [review]
Thanks for the time you've spent on this module, it's very much appreciated.

I've not looked too closely at the rest of the code, this review is specifically with regards the boto3/botocore requirements.

Please don't adjust the *global* botocore/boto3 requirements.

The current collection policy is that we support the version of boto3/botocore released about 1 year prior to our current major release (currently botocore 1.19/boto3 1.16).  While folks wanting support for a new AWS may be happy constantly updating botocore/boto3 this can be frustrating for folks who find that their previously working environment stops working after a minor/patch release of the collection.

Individual modules, and deliberately fenced off features inside other modules can use newer versions of the AWS SDKs, however, they should be explicitly testing for them.  Otherwise we **do** get issues/complaints caused by modules suddenly no longer working because we added support for a new feature they didn't care about.  Unfortunately unless you're familiar with the types of errors you'll see it's also non-trivial to diagnose the issue (and non-trivial to automatically catch too).

Manually catching exactly which version a botocore feature was introduced generally doesn't work and as such, to ensure we don't accidentally depend on features from a later release, our tests are deliberately run against the minimum 'supported' version.

Tests for modules/features that need newer versions should themselves by testing against the minimum necessary botocore versions.  This can be installed by using the helper role "setup_botocore_pip" and used by defining the `ansible_python_interpreter`.  See ec2_asg_scheduled_action for an example.  This mechanism is, unfortunately, currently poorly documented and only needed by a handful of modules, which we should fix.

File: plugins/modules/opensearch.py (hunk 3 above)
Comment (alinabuzachis) [reply plugins/modules/opensearch.py:575]: I suggested to move the helper methods in a new opensearch.py file here  https://github.com/ansible-collections/community.aws/tree/main/plugins/module_utils. Then, you can import them in your module (no need to open the PR in amazon.aws).

@tremble  What do you think about moving `domain to_facts()` -  maybe we can rename this method to get_tags for example, `ensure_tags()` and `get_domain()` inside https://github.com/ansible-collections/community.aws/tree/main/plugins/module_utils?

Comment (tremble) [reply plugins/modules/opensearch.py:575]: @sebastien-rosset

Unfortunately, we don't have a lot of historical consistency in the way our modules have generally been written, so I ask for your understanding when it comes to "but other modules don't do that".  Cleaning up is a massive job, and is getting picked at very slowly.

Helper functions that can be used by other services should live in `amazon.aws`.  However, anything OpenSearch specific should live in `community.aws` unless or until the modules themselves move to `amazon.aws`.  Personally I'd stick to only making changes here in `community.aws` for now, we can talk about refactoring things over into amazon.aws later if appropriate.  The kinds of OpenSearch specific helpers @alinabuzachis's talking about would move into `plugins/module_utils/opensearch.py`.

FWIW, some modules ended up being split across amazon.aws and community.aws initially and it's been more painful to test and manage than was predicted (there was a reasonable logic behind it, but "hindsight").

You don't see many modules using module_utils today because things have grown organically and it was "complicated" prior to the Ansible- Core/Collections split.  However, the some_service/some_service_info model has resulted in far more copy and paste than it should because people don't use module_utils.  Starting cleanly by moving generic code into module_utils/ is technically the better idea.

Specifically with regards "XYZ_to_facts" as a name, we almost never return **facts** from our modules these days.  I tend to refer to is as "normalize_XYZ", but it's a good example of a piece of code best handled in module_utils, because it'll be needed/wanted by both opensearch and a potential future opensearch_info module.

Comment (sebastien-rosset) [reply plugins/modules/opensearch.py:575]: Sure, no problem I understand the challenges of history. If I go on the path of splitting the module to externalize functions to `module_utils`, I need to find a way to use the `opensearch` module in a environment until `community.aws` version 3.1 is released. Reviewing PRs, releasing software and testing the new software version usually takes weeks or months, meanwhile I would like to be able to use the `opensearch` module for our needs.

When the module is self contained as a single file, the transition is very easy: modify `ansible.cfg`, add a `library = ./library` statement, drop a copy of `opensearch.py` under `./library`, and that's it. We usually commit the new or patched module to our repos. Then later when changes have been upstreamed, released and tested, we can upgrade the version of the ansible package and remove the locally patched module.

When the module is not self contained, the transition period looks more complicated. I hope that does not mean we have to build the `community.aws` fork and install it in every developer environment as a replacement of the upstream version.

Comment (sebastien-rosset) [reply plugins/modules/opensearch.py:520]: Is this what you had in mind? b668e01ad381b54c7299b395e44c1674020005b6

Comment (alinabuzachis) [reply plugins/modules/opensearch.py:538]: @sebastien-rosset yes, exactly! You should also move `ensure_tags()` to module_utils. Thanks.

Comment (sebastien-rosset) [reply plugins/modules/opensearch.py]: done

**Comment by gravesm** [comment]
@sebastien-rosset Thank you for your work on this. While we usually love to see integration tests accompany PRs, our CI is not equipped to handle a 4+ hour test suite. Even if we made the necessary policy changes, this PR will fail due to timeout. In this case, we should try and test whatever we can through unit tests.

**Comment by sebastien-rosset** [comment]
The integration tests were useful to develop the module. How about retaining the integration tests but disabling them? That way at least it's possible to run these tests in a local workspace.

**Comment by tremble** [comment]
I'm a fan of keeping them, but disabling them.  Even if they won't be run by CI, we can always run them manually if they're there.

It would be really good if one of the bots was aware that the tests that should theoretically run, but are disabled / unstable.  I wouldn't know where to start with that though.

**Comment by markuman** [comment]
I run locally into this. 

```
ansible-test integration -v --python 3.7 opensearch --docker

...


TASK [opensearch : include_tasks] **********************************************
fatal: [testhost]: FAILED! => {"reason": "couldn't resolve module/action 'community.crypto.openssl_privatekey'. This often indicates a misspelling, missing collection, or incorrect module path.\n\nThe error appears to be in '/root/ansible_collections/community/aws/tests/output/.tmp/integration/opensearch-vty8s_7c-ÅÑŚÌβŁÈ/tests/integration/targets/opensearch/tasks/test_create_cert.yml': line 13, column 3, but may\nbe elsewhere in the file depending on the exact syntax problem.\n\nThe offending line appears to be:\n\n  register: tempdir_1\n- name: Generate private key\n  ^ here\n"}
[WARNING]: Failure using method (v2_runner_on_failed) in callback plugin
(<ansible.plugins.callback.junit.CallbackModule object at 0x7f762b8fad50>): /ro
ot/ansible_collections/community/aws/tests/output/.tmp/integration/opensearch-
vty8s_7c-ÅÑŚÌβŁÈ/tests/integration/targets/opensearch/tasks/main.yml:26:
testhost: opensearch : include_tasks _raw_params=test_create_cert.yml:
duplicate host callback: testhost

```

**Comment by sebastien-rosset** [comment]
> couldn't resolve module/action 'community.crypto.openssl_privatekey
What version of `community.crypto` do you have?
Mine is:
community.crypto              2.0.1

**Comment by markuman** [comment]
> What version of `community.crypto` do you have?
> Mine is:
> community.crypto 2.0.1

 2.2.2 and the module exists since 1.0.0  
but local version doesn't matter when testing via docker environment.   
furthermore, the module is also used in other integration tests and the collection itself is defined in `tests/requirements.yml`.  
I'm running out of ideas why it is failing atm.

File: tests/integration/targets/opensearch/aliases [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+#   35 minutes to upgrade cluster from 7.1 to 7.10.
+#   23 minutes to enable node-to-node encryption.
+#   36 minutes to enable encryption at rest.
+#   30 minutes to enable warm storage.
+#   30 minutes to enable cold storage.
+#   30 minutes to enable and enforce HTTPS on the domain endpoint.
+#    3 minutes to enable auto-tune option.
+#   45 minutes to delete cluster.
+
+#disabled
```
Comment (sebastien-rosset) [inline tests/integration/targets/opensearch/aliases:22]:
```suggestion
disabled
```

**Comment by sebastien-rosset** [comment]
> I'm a fan of keeping them, but disabling them. Even if they won't be run by CI, we can always run them manually if they're there.
> 
> It would be really good if one of the bots was aware that the tests that should theoretically run, but are disabled / unstable. I wouldn't know where to start with that though.

Based on your input, should I apply this change?
https://github.com/ansible-collections/community.aws/pull/859/files#diff-f02aaab97cdd3c1cff0784c53b3bb82b056e9bd456b8d10fbb337f9a7fbd1c23R22

File: tests/integration/targets/opensearch/aliases (hunk 11 above)
Comment (alinabuzachis) [reply tests/integration/targets/opensearch/aliases:22]: That’s correct.

Comment (alinabuzachis) [reply tests/integration/targets/opensearch/aliases:22]: We run the tests locally and disable them in the CI.

**Review by alinabuzachis — APPROVED** [review]
@sebastien-rosset Thank you. LGTM!

File: tests/integration/targets/opensearch/tasks/test_vpc_setup.yml [hunk 12]
```diff
@@ -0,0 +1,134 @@
...
+          Name: "{{ tiny_prefix }}-zone"
+          AnsibleTest: AnsibleTestVpc
+
+    - name: Set fact for route53 zone id
+      set_fact:
+        route53_zone_id: "{{ route53_zone.result.zone_id }}"
+
+    - name: Create KMS key for test purpose
+      # The key is needed for OpenSearch encryption at rest.
+      aws_kms:
```
Comment (markuman) [inline tests/integration/targets/opensearch/tasks/test_vpc_setup.yml:123]: last thing.  
the aws_kms key should also be cleaned up.

Comment (sebastien-rosset) [reply tests/integration/targets/opensearch/tasks/test_vpc_setup.yml:123]: This is already done here: https://github.com/ansible-collections/community.aws/pull/859/files#diff-0a2082f73b60bcc897b26905fabe43ec4ab77ced8f016f27014ae7e3bbbf8078R45-R56

**Comment by alinabuzachis** [comment]
recheck

**Comment by sebastien-rosset** [comment]
Is there a chance this will be reviewed?

File: plugins/modules/opensearch.py [hunk 13]
```diff
@@ -0,0 +1,1507 @@
...
+__metaclass__ = type
+
+
+DOCUMENTATION = """
+---
+module: opensearch
+short_description: Creates OpenSearch or ElasticSearch domain.
+description:
+  - Creates or modify a Amazon OpenSearch Service domain.
+version_added: 3.1.0
```
Comment (tremble) [inline plugins/modules/opensearch.py:16]:
```suggestion
version_added: 4.0.0
```

File: plugins/modules/opensearch_info.py [hunk 14]
```diff
@@ -0,0 +1,530 @@
...
+__metaclass__ = type
+
+
+DOCUMENTATION = """
+---
+module: opensearch_info
+short_description: obtain information about one or more OpenSearch or ElasticSearch domain.
+description:
+  - obtain information about one Amazon OpenSearch Service domain.
+version_added: 3.1.0
```
Comment (tremble) [inline plugins/modules/opensearch_info.py:16]:
```suggestion
version_added: 4.0.0
```
